p2p/discover/v5_encoding.go

```diff
@@ -0,0 +1,649 @@
+		}
+		return enc, nil, err
+	}
+	// Ensure calling code sets node if needed.
+	if challenge != nil && challenge.node == nil {
+		panic("BUG: missing challenge.node in encode")
+	}
+	writeKey := c.sc.writeKey(id, addr)
+	if writeKey != nil || challenge != nil {
+		return c.encodeEncrypted(id, addr, packet, writeKey, challenge)
+	}
+	return c.encodeRandom(id)
+}
+
+// encodeRandom encodes a random packet.
+func (c *wireCodec) encodeRandom(toID enode.ID) ([]byte, []byte, error) {
+	tag := xorTag(c.sha256sum(toID[:]), c.localnode.ID())
+	r := make([]byte, 44) // TODO randomize size
+	if _, err := crand.Read(r); err != nil {
+		return nil, nil, err
+	}
+	nonce := make([]byte, gcmNonceSize)
+	if _, err := crand.Read(nonce); err != nil {
+		return nil, nil, fmt.Errorf("can't get random data: %v", err)
+	}
+	b := new(bytes.Buffer)
+	b.Write(tag[:])
+	rlp.Encode(b, nonce)
+	b.Write(r)
+	return b.Bytes(), nonce, nil
+}
+
+// encodeWhoareyou encodes WHOAREYOU.
+func (c *wireCodec) encodeWhoareyou(toID enode.ID, packet *whoareyouV5) ([]byte, error) {
+	// Sanity check node field to catch misbehaving callers.
+	if packet.RecordSeq > 0 && packet.node == nil {
+		panic("BUG: missing node in whoareyouV5 with non-zero seq")
+	}
+	b := new(bytes.Buffer)
+	b.Write(c.sha256sum(toID[:], []byte("WHOAREYOU")))
+	err := rlp.Encode(b, packet)
+	return b.Bytes(), err
+}
+
+// encodeEncrypted encodes an encrypted packet.
+func (c *wireCodec) encodeEncrypted(toID enode.ID, toAddr string, packet packetV5, writeKey []byte, challenge *whoareyouV5) (enc []byte, authTag []byte, err error) {
+	nonce := make([]byte, gcmNonceSize)
+	if _, err := crand.Read(nonce); err != nil {
+		return nil, nil, fmt.Errorf("can't get random data: %v", err)
+	}
+
+	var headEnc []byte
+	if challenge == nil {
+		// Regular packet, use existing key and simply encode nonce.
+		headEnc, _ = rlp.EncodeToBytes(nonce)
+	} else {
+		// We're answering WHOAREYOU, generate new keys and encrypt with those.
+		header, sec, err := c.makeAuthHeader(nonce, challenge)
+		if err != nil {
+			return nil, nil, err
+		}
+		if headEnc, err = rlp.EncodeToBytes(header); err != nil {
+			return nil, nil, err
+		}
+		c.sc.storeNewSession(toID, toAddr, sec.readKey, sec.writeKey)
+		writeKey = sec.writeKey
+	}
+
+	// Encode the packet.
+	body := new(bytes.Buffer)
+	body.WriteByte(packet.kind())
+	if err := rlp.Encode(body, packet); err != nil {
+		return nil, nil, err
+	}
+	tag := xorTag(c.sha256sum(toID[:]), c.localnode.ID())
+	headsize := len(tag) + len(headEnc)
+	headbuf := make([]byte, headsize)
+	copy(headbuf[:], tag[:])
+	copy(headbuf[len(tag):], headEnc)
+
+	// Encrypt the body.
+	enc, err = encryptGCM(headbuf, writeKey, nonce, body.Bytes(), tag[:])
+	return enc, nonce, err
+}
+
+// encodeAuthHeader creates the auth header on a call packet following WHOAREYOU.
+func (c *wireCodec) makeAuthHeader(nonce []byte, challenge *whoareyouV5) (*authHeaderList, *handshakeSecrets, error) {
```

Comment:
I should really add some comments to this function, it's the most complicated part